### ISSUE TITLE
Interceptor Timeout

### ISSUE THREAD

**Issue by dibyom** [tektoncd/triggers#1452]
### Discussed in https://github.com/tektoncd/triggers/discussions/1451

<div type='discussions-op-text'>

<sup>Originally posted by **joshua-blickensdoerfer** September 29, 2022</sup>
Hello,

i have written a custom Cluster interceptor. In the eventlistener log i can see that there is a timeout after a few seconds

"logger":"eventlistener","caller":"sink/sink.go:381","msg":"Post \"http://custom-svc.tekton-framework.svc:80\": net/http: timeout awaiting response headers"

Is there any way to increase the timeout duration for custom interceptors? I've tried increasing the value of "'-el-readtimeout" and "-el-httpclient-readtimeout" this did not seem to have any effect.

Kind regards
Joshua</div>

**Comment by jmcshane** [comment]
Right now, the way this could be done on the eventlistener would be to add custom args for overriding the defaults for the http client. We could build an HTTP client for each cluster interceptor, this could simplify the default EL http client construction as we have to assemble the full tls config for all the interceptors on startup right now in https://github.com/tektoncd/triggers/blob/v0.21.0/pkg/adapter/adapter.go#L124 and keep a watch on it to continually update.

I could see a clusterinterceptorspec like:

```
kind: ClusterInterceptor
...
spec:
  timeouts:
    tlshandshake:
    responseheader:
    expectcontinuetimeout: 
    readtimeout:
    keepalive:
```

Obviously, these could all be optional values so we can distinguish being unset vs set to 0, but I'm thinking about the "default" behavior here. Would `nil` mean "default to the current eventlistener value" vs `0` meaning "no timeout"? Are we concerned about the penalty for rebuilding the interceptor httpclient on every interceptor call?

**Comment by dibyom** [comment]
> Would nil mean "default to the current eventlistener value" vs 0 meaning "no timeout"?

Yeah I think that makes sense

>  Are we concerned about the penalty for rebuilding the interceptor httpclient on every interceptor call?

I think so 😬 do we need to build the interceptor on each call? can we do it periodically or when needed if a interceptor changes?

(Doesn't help with timeouts but for certs at least we could provide tls.Config's GetCertificate similar to how knative/pkg's [webhook implementation does:](https://github.com/knative/pkg/blob/main/webhook/webhook.go#L145-L167))

**Comment by jmcshane** [comment]
> I think so 😬 do we need to build the interceptor on each call? can we do it periodically or when needed if an interceptor changes?

Yeah, that was my presumption as well. Let me take a look at how the interceptor watch works and see if we can keep these clients somewhere reasonable and just update on watches

**Comment by khrm** [comment]
/remove-lifecycle rotten

/lifecycle-frozen

We will handle this in future releases.

**Comment by keskad** [comment]
I wrote an interceptor that fetches missing information from Gitlab, when a webhook is missing them, so it could decide if the Pipeline should be started or not. The issue is I suppose when Gitlab timeouts, then my interceptor also timeouts, then the Event Listener shows "http: server closed idle connection", but I'm not 100% sure. I was thinking about increasing read timeout, but I get validation error when trying to add commandline args in Event Listener object:

```
one or more objects failed to apply, reason: error when patching "/dev/shm/1121367218": admission webhook "validation.webhook.triggers.tekton.dev" denied the request: validation failed: must not set the field(s): spec.resources.kubernetesResource.spec.template.spec.containers[0].args (retried 5 times). 
```
